src/utils.c

```diff
@@ -36,29 +36,50 @@ bool allNA(SEXP x) {
   // less space and time than any(is.na(x)) at R level because that creates full size is.na(x) first before any()
   // whereas this allNA can often return early on testing the first value without reading the rest
   const int n = length(x);
-  if (n==0) return true; // for empty raw, and NULL, to match R's all(is.na(raw())) true result, test 2116.13-14
+  if (n==0) // for empty raw, and NULL, to match R's all(is.na(raw())) true result, test 2116.13-14
+    return true;
   switch (TYPEOF(x)) {
   case LGLSXP:
   case INTSXP: {
     const int *xd = INTEGER(x);
-    for (int i=0; i<n; ++i)   if (xd[i]!=NA_INTEGER)   return false;
+    for (int i=0; i<n; ++i) {
+      if (xd[i]!=NA_INTEGER)
+        return false;
+    }
     return true;
   }
   case REALSXP:
     if (Rinherits(x,char_integer64)) {
       const int64_t *xd = (int64_t *)REAL(x);
-      for (int i=0; i<n; ++i) if (xd[i]!=NA_INTEGER64) return false;
+      for (int i=0; i<n; ++i) {
+        if (xd[i]!=NA_INTEGER64)
+          return false;
+      }
     } else {
       const double *xd = REAL(x);
-      for (int i=0; i<n; ++i) if (!ISNAN(xd[i]))       return false;
+      for (int i=0; i<n; ++i) {
+        if (!ISNAN(xd[i]))
+          return false;
+      }
     }
     return true;
   case STRSXP: {
     const SEXP *xd = STRING_PTR(x);
-    for (int i=0; i<n; ++i)   if (xd[i]!=NA_STRING)    return false;
+    for (int i=0; i<n; ++i) {
+      if (xd[i]!=NA_STRING)
+        return false;
+    }
     return true;
-  }}
-  return false;
+  }
+  case VECSXP: {
+    return false;
+  }
+  case RAWSXP: {
+    return false;
+  }
+  default:
+    error("Internal error: unsupported type '%s' passed to allNA()", type2char(TYPEOF(x))); // # nocov
```

Comment:
We have `raw` and `list` in tests, instead of returning `false` by default we return `false` only for those two, and raise error otherwise. I still see a potential for issues here, for example:
If user defines own class using a list type. Then user can have `is.na.my_class` method, which can work on list, while our `allNA` will not work returning FALSE always.
```
my_cls = structure(list(a=1), class="my_class")
is.na.my_class = function(x) TRUE
```

Reply:
Good points. `raw` doesn't support NA so that's why `raw`, other than `allNA(raw())` is true and I included test for.  All empty vectors return true.
I wasn't planning on exporting `allNA` in this release.  Just using it internally to start with.
`list`: yes there's some wrinkles there.  `allNA(list())` is correctly true but I don't think I covered a test for that.  But `allNA(list(NA))` is false whereas `all(is.na(list(NA)))` is true in base R. So that needs fixing.  I don't think we pass `list` to `allNA` internally in this PR,  but it should be error at least until it supports it,  otherwise we're just relying on test coverage too much.  Will do.

Reply:
base R behaviour:
```R
> all(is.na(list(NA, NA)))
[1] TRUE
> all(is.na(list(NA, c(NA))))
[1] TRUE
> all(is.na(list(NA, c(NA,NA))))
[1] FALSE
```

Reply:
there's an explicit

```
	if (!isVector(s) || length(s) != 1)
	    pa[i] = 0;
```

in `do_isna` in `src/main/coerce.c`. No comment as to why... not changed in 19 years

https://github.com/wch/r-source/commit/74bb7a0bdc5e4b21464ec87343f35bf834a6620c

Reply:
Interesting.  Actually, turning off list support in allNA (so that it errors with unsupported) does cause 3 tests to fail:  1294.05, 1294.12 and 1294.19.   Just in this PR because this PR is now using allNA in bmerge.

Reply:
And those 3 tests all warn with `Coerced list RHS to logical|integer|double` with on RHS `list(list(1))`.  Comforting that the circle leads back to this `coerce_01` PR.

Reply:
Since those 3 were warnings already, and with #3626 in mind, I turned them to error and moved on.  It's strange we were allowing these `list` singletons to be coerced to atomic, albeit with warning.  I can see it causing problems with #3626 so those 3 being error make sense for now I hope.